docs/devel/infrastructure-interaction.md

```diff
@@ -0,0 +1,85 @@
+
+If we attack the problem right now, today, as problem #3 it also leaves the door open for people to write tools on top of kube and their infrastructure to mimic #1 and #2.  Which can hopefully help us one day bring those tools into kube.  But decoupling is a very important first step.
+
+Between minions, GCE PDs, and GCE external load balancers, all of these have different implementations with little coherence for non-gce configurations. So thinking generally, how should kube and infrastructure functionality work.  If I’m looking at the kube cluster from that 3rd infrastructure interaction perspective (and possibly even the 1st perspective) I think we must view that underlying functionality as pets, not cattle.  That’s just a perspective I’m bringing into this discussion.
+
+To decouple kube and the infrastructure I believe that our model of interaction should require three distinct steps and three distinct parties interacting.
+
+**Design Idea**
+
+There are three parties who need to operate on the system in my mind.  They may be actual people running things on a command line or automated tools doing something similar.  That’s not really relevant.
+
+1. Infrastructure admin.  The person with the permission and ability to make changes to the infrastructure.
+2. Kube admin.  The person with the permission to change how the kube cluster itself operates.
+3. Kube user.  The person who just wants to write and run an app.
+
+My actual design decoupling can be broken down into 3 steps.
+
+1. The “infrastructure admin”, needs to make these external functions available to kube, potentially via a cloud provider interface similar to how minions are handled today.  We would want something similar for storage and external reachability functionality.  It seems likely that we may want to decompose the cloud_provider into multiple smaller providers.  ‘minion_provider’, ‘storage_provider’, etc.  But that is a specific detail to work out per infrastructure concept.
+
+2. The “kube admin” needs to explicitly tell the cluster to use the functionality exposed by the infrastructure. In the case of minions this defaults to “*” but we have the –minion_regexp. That means minions sorta fit this model. GCE PDs fit as you have to specify the GCEPersistentDisk.  The specifics of how the kube admin tells the cluster to use the functionality will likely be functionality specific.  (Although I’ll give my thoughts on some of these things below)
```

Comment:
You're describing GCEPersistentDisk as if it were a plugin, or you just consider it an admin choice regarding where the cluster is run? Either way, I suppose it makes sense.